internal/storage/field_stats.go

```diff
@@ -0,0 +1,355 @@
+		stats.Max = pk
+	} else if stats.Max.LT(pk) {
+		stats.Max = pk
+	}
+}
+
+func NewFieldStats(fieldID int64, pkType schemapb.DataType, rowNum int64) (*FieldStats, error) {
+	return &FieldStats{
+		FieldID: fieldID,
+		Type:    pkType,
+		BF:      bloom.NewWithEstimates(uint(rowNum), paramtable.Get().CommonCfg.MaxBloomFalsePositive.GetAsFloat()),
+	}, nil
+}
+
+// FieldStatsWriter writes stats to buffer
+type FieldStatsWriter struct {
+	buffer []byte
+}
+
+// GetBuffer returns buffer
+func (sw *FieldStatsWriter) GetBuffer() []byte {
+	return sw.buffer
+}
+
+// GenerateList writes Stats slice to buffer
+func (sw *FieldStatsWriter) GenerateList(stats []*FieldStats) error {
+	b, err := json.Marshal(stats)
+	if err != nil {
+		return err
+	}
+	sw.buffer = b
+	return nil
+}
+
+// GenerateByData writes Int64Stats or StringStats from @msgs with @fieldID to @buffer
+func (sw *FieldStatsWriter) GenerateByData(fieldID int64, pkType schemapb.DataType, msgs ...FieldData) error {
+	statsList := make([]*FieldStats, 0)
+	for _, msg := range msgs {
+		stats := &FieldStats{
+			FieldID: fieldID,
+			Type:    pkType,
+			BF:      bloom.NewWithEstimates(uint(msg.RowNum()), paramtable.Get().CommonCfg.MaxBloomFalsePositive.GetAsFloat()),
+		}
+
+		stats.UpdateByMsgs(msg)
+		statsList = append(statsList, stats)
+	}
+	return sw.GenerateList(statsList)
+}
+
+// FieldStatsReader reads stats
+type FieldStatsReader struct {
+	buffer []byte
+}
+
+// SetBuffer sets buffer
+func (sr *FieldStatsReader) SetBuffer(buffer []byte) {
+	sr.buffer = buffer
+}
+
+// GetFieldStats returns buffer as FieldStats
+func (sr *FieldStatsReader) GetFieldStats() (*FieldStats, error) {
+	stats := &FieldStats{}
+	err := json.Unmarshal(sr.buffer, &stats)
+	if err != nil {
+		return nil, merr.WrapErrParameterInvalid("valid JSON", string(sr.buffer), err.Error())
+	}
+
+	return stats, nil
+}
+
+// GetFieldStatsList returns buffer as FieldStats
+func (sr *FieldStatsReader) GetFieldStatsList() ([]*FieldStats, error) {
+	var stats []*FieldStats
+	err := json.Unmarshal(sr.buffer, &stats)
+	if err != nil {
+		return nil, merr.WrapErrParameterInvalid("valid JSON", string(sr.buffer), err.Error())
+	}
+
+	return stats, nil
+}
+
+// DeserializeFieldStats deserialize @blobs as []*FieldStats
+func DeserializeFieldStats(blobs []*Blob) ([]*FieldStats, error) {
+	results := make([]*FieldStats, 0, len(blobs))
+	for _, blob := range blobs {
+		if len(blob.Value) == 0 {
+			continue
+		}
+		sr := &FieldStatsReader{}
+		sr.SetBuffer(blob.Value)
+		stats, err := sr.GetFieldStats()
+		if err != nil {
+			return nil, err
+		}
+		results = append(results, stats)
+	}
+	return results, nil
+}
+
+func DeserializeFieldStatsList(blob *Blob) ([]*FieldStats, error) {
+	if len(blob.Value) == 0 {
+		return []*FieldStats{}, nil
+	}
+	sr := &FieldStatsReader{}
+	sr.SetBuffer(blob.Value)
+	stats, err := sr.GetFieldStatsList()
+	if err != nil {
+		return nil, err
+	}
+	return stats, nil
+}
+
+type PartitionStats struct {
```

Comment:
Name it PartitionStatsSnapshot?

Since partition is mutable,  the version actually represent the state of this snapshot?

when we put this on S3,
path should be prefix/collectionID/partitionID/version